File: mean_shift/mean_shift.R
library(ggplot2)
library(meanShiftR)

# Read data
data("iris")

# Explore data
ggplot(iris, aes(Petal.Length, Petal.Width, color = Species)) +
  geom_point() + xlab("Petal length") + ylab ("Petal width")

petals <- cbind(iris$Petal.Length, iris$Petal.Width)

# Train the model
tags <- meanShift(queryData = petals, trainData = petals, 
                  bandwidth = c(0.7, 0.7))
iris$tags <- as.factor(tags$assignment)

# Explore the model
ggplot(iris, aes(Petal.Length, Petal.Width, color = tags)) +
  geom_point() + xlab("Petal length") + ylab ("Petal width")

MC <- table(iris$tags, iris$Species)

# Get model accuracy
sum(diag(MC))/sum(MC)

# Draw it
centroids <- unique(tags$value)
ggplot(iris, aes(Petal.Length, Petal.Width, color = tags)) +
  geom_point() + xlab("Petal length") + ylab ("Petal width") +
  geom_point(aes(x= centroids[1,1], y= centroids[1,2]),
             colour="red",shape = 24)+
  geom_point(aes(x= centroids[2,1], y= centroids[2,2]),
             colour="green",shape = 24)+
  geom_point(aes(x= centroids[3,1], y= centroids[3,2]),
             colour="blue",shape = 24)

File: mean_shift/mean_shift.py
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.datasets import load_iris
from sklearn.cluster import MeanShift

# Read data
iris_data = load_iris()
iris = load_iris()
irisData = pd.DataFrame(data = iris.data)
irisData['Species'] = iris.target

# Explore data
# We will only use Petal length and width
plt.scatter(irisData[[2]], irisData[[3]], c = irisData['Species'])
plt.xlabel("Petal length")
plt.ylabel("Petal width")
plt.show()

#exit()

# Train and test the model
petals = irisData[[2, 3]].values
model = MeanShift(bandwidth = 0.7).fit(petals)
tags = model.labels_
irisData['tags'] = tags

# Explore model
plt.scatter(irisData[[2]], irisData[[3]], c = irisData['tags'])
plt.xlabel("Petal length")
plt.ylabel("Petal width")
plt.show()

#exit()

# Get model accuracy
MC = pd.crosstab(irisData['tags'], irisData['Species'])
accuracy = sum(MC.values.diagonal()) / MC.values.sum()
print("Model accuracy:", accuracy)

# Draw it
centroids = model.cluster_centers_
plt.scatter(irisData[[2]], irisData[[3]], c = irisData['tags'])
plt.scatter(centroids[:, 0], centroids[:, 1], c = ['red', 'green', 'blue'], marker = 'o', s = 100)
plt.xlabel("Petal length")
plt.ylabel("Petal width")
plt.show()
